docs/replica-sets.md

```diff
@@ -0,0 +1,117 @@
+# VirtualMachineReplicaSet
+
+## Overview
+
+In order to allow scaling up or down similar VMs, add a way to specify
+`VirtualMachine` templates and the amount of replicas required, to let the
+runtime create these `VirtualMachine`s.
+
+## Use-cases
+
+### Ephemeral VMs
+
+Scaling ephemeral VMs which only have read-only mounts, or work with a backing
+store, to keep temporary data, which can be deleted after the VM gets
+undefined.
+
+## Implementation
+
+A new object `VirtualMachineReplicaSet`, backed with a controller will be
+implemented:
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachineReplicaSet
+metadata:
+  name: myreplicaset
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      mylabel: mylabel
+  template:
+    metadata:
+      name: test
+      labels:
+        mylabel: mylabel
+    spec:
+      domain:
+        devices:
+      [...]  
+```
+
+`spec.template` is equal to a `VirtualMachineSpec`. `spec.replicas` specifies
+how many instances should be created out of `spec.temlate`. `spec.selector`
+contains selectors, which need to match `spec.template.metadata.labels`.
+
+The status looks like this:
+
+```yaml
+status:
+  conditions: null
+  replicas: 3
+```
+
+It shows the number of `VirtualMachine`s which are in a non-final state nad
+which match `spec.selector`. In case of a scaling error, a `ReplicaFailure`
+condition is added to the status.
+
+In case of a delete failure:
+
+```yaml
+status:
+  conditions:
+  - type: "ReplicaFailure"
+    status: True
+    reason: "FailureDelete"
+    message: "no permission to delete VMs"
+    lastTransmissionTime: "..."
+  replicas: 4
+```
+
+In case of a create failure:
+
+```yaml
+status:
+  conditions:
+  - type: "ReplicaFailure"
+    status: True
+    reason: "FailureCreate"
+    message: "no permission to create VMs"
+    lastTransmissionTime: "..."
+  replicas: 2
+```
+
+### Guarantees
+
+Once Graceful Deletes are implemented, the VirtualMachineReplicaSet guarantees
+that it will never create more than the requested numbers of VMs in a non-final
```

Comment:
I think, that as soon as we have graceful delete (which means that only the administrator or virt-handler would do the final delete) we can give that guarantee. However, if no fencing action is taken and VMs start piling up in unknown or pending states, the replication controller will never create new VMs until the situation is resolved.

Reply:
Given that this is not a guarantee that ReplicaSets give it would be better to pick a different name.  A replica set is explictly created to mint fungible copies.  This guarantee is not part of what replicas do.

Reply:
Ok, good point. Will remove that guarantee.